### PR TITLE
[CELEBORN-1472] Reduce  CongestionController#userBufferStatuses call times.

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
Reduce  CongestionController#userBufferStatuses call times.

### Why are the changes needed?
When we use sort based shuffle writer, The number of PushMergedData requests has increased which make CongestionController#produceBytes taking up much cpu time.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Existing uts.

### CODE REVIEW

**Comment by leixm** [comment]
WechatIMG89

**Comment by leixm** [comment]
@AngersZhuuuu  @pan3793  @waitinfuture  PTAL.

File: worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala [hunk 1]
```diff
@@ -1282,6 +1282,8 @@ class PushDataHandler(val workerSource: WorkerSource) extends BaseMessageHandler
           fileWriter.decrementPendingWrites()
       }
     }
+
+    updateBytesProduced(fileWriters.head, body.readableBytes())
```
[the remark this answers is not in the document]
Comment (leixm) [reply worker/src/main/scala/org/apache/celeborn/service/deploy/worker/PushDataHandler.scala]: You are right, I will modify another place, it seems that there is no need to call `userBufferStatuses.computeIfAbsent ` every time

**Comment by leixm** [comment]
Found through stress testing, we reduced the CPU usage of produceBytes from **25.07% to 16.58%**
Before this PR
kk1

After this PR
kk2

**Comment by AngersZhuuuu** [comment]
Pls update pr title and pr desc

**Comment by leixm** [comment]
@AngersZhuuuu  @waitinfuture  PTAL.

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/congestcontrol/CongestionController.java [hunk 2]
```diff
@@ -275,4 +252,19 @@ public static synchronized void destroy() {
...
+          logger.info("New user {} comes, initializing its rate status", user);
+          BufferStatusHub bufferStatusHub = new BufferStatusHub(sampleTimeWindowSeconds);
+          UserBufferInfo userInfo = new UserBufferInfo(System.currentTimeMillis(), bufferStatusHub);
+          workerSource.addGauge(
+              WorkerSource.USER_PRODUCE_SPEED(),
+              userIdentifier.toJMap(),
+              () -> getUserProduceSpeed(userInfo));
+          return userInfo;
+        });
+  }
```
Comment (AngersZhuuuu) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/congestcontrol/CongestionController.java:269]: Pls not put this method at the tail of this class code

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 3]
```diff
@@ -288,10 +295,10 @@ public void write(ByteBuf data) throws IOException {
...
     } else {
       MemoryManager.instance().incrementDiskBuffer(numBytes);
-      Optional.ofNullable(CongestionController.instance())
-          .ifPresent(
-              congestionController ->
-                  congestionController.produceBytes(diskFileInfo.getUserIdentifier(), numBytes));
+      if (userBufferInfo != null) {
+        userBufferInfo.updateInfo(
+            System.currentTimeMillis(), new BufferStatusHub.BufferStatusNode(numBytes));
+      }
```
Comment (AngersZhuuuu) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:301]: We can remove `congestionController.produceBytes()`

**Review by AngersZhuuuu — APPROVED** [review]
LGTM

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 4]
```diff
@@ -153,6 +156,10 @@ public PartitionDataWriter(
       this.mapIdBitMap = new RoaringBitmap();
     }
     takeBuffer();
+    CongestionController congestionController = CongestionController.instance();
+    if (!isMemoryShuffleFile.get() && congestionController != null) {
+      userBufferInfo = congestionController.getUserBuffer(getDiskFileInfo().getUserIdentifier());
```
Comment (RexXiong) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:161]: IMO the CongestionController should control all file write operations including memory file ?

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java [hunk 5]
```diff
@@ -288,10 +295,10 @@ public void write(ByteBuf data) throws IOException {
       MemoryManager.instance().increaseMemoryFileStorage(numBytes);
     } else {
       MemoryManager.instance().incrementDiskBuffer(numBytes);
-      Optional.ofNullable(CongestionController.instance())
-          .ifPresent(
-              congestionController ->
-                  congestionController.produceBytes(diskFileInfo.getUserIdentifier(), numBytes));
+      if (userBufferInfo != null) {
+        userBufferInfo.updateInfo(
```
Comment (RexXiong) [inline worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:299]: ditto

File: worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java (hunk 4 above)
Comment (leixm) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:161]: I keep  the original code logic unchanged. In the original logic, CongestionController does not manage memory files. Maybe I can correct this logic in the next PR?

Comment (RexXiong) [reply worker/src/main/java/org/apache/celeborn/service/deploy/worker/storage/PartitionDataWriter.java:161]: > I keep the original code logic unchanged. In the original logic, CongestionController does not manage memory files. Maybe I can correct this logic in the next PR?

LGTM

**Review by RexXiong — APPROVED** [review]

**Comment by RexXiong** [comment]
Merge to main(v0.6.0) and branch-0.5(v0.5.1)

**Comment by leixm** [comment]
Thank you for you review. @AngersZhuuuu  @RexXiong

**Pull request closed without merging**